python/ray/autoscaler/_private/aws/events.py

```diff
@@ -0,0 +1,140 @@
+        raise NotImplementedError("AWS Lambda callback is not implemented")
+
+    def _cloudwatch_callback(self):
+        raise NotImplementedError("AWS Cloudwatch callback is not implemented")
+
+    def _api_gateway_callback(self):
+        raise NotImplementedError("AWS API Gateway callback is not implemented")
+
+
+class AwsEventManager(AwsEventManagerBase):
+    def __init__(self, events_config: Dict[str, Any]):
+        self.events_config = events_config
+        self.uri = notification_uri = events_config["notification_uri"]
+        assert notification_uri is not None, f"`notification_uri` is a required field in `events`"
+        assert "arn:aws" in notification_uri, f"Invalid ARN specified: {notification_uri}"
+        self.parameters = events_config.get("parameters", {})
+
+    @property
+    def config(self) -> Dict[str, Any]:
+        return self.events_config
+
+    @property
+    def metadata(self) -> Dict[str, Any]:
+        return self.parameters
+
+    def _sns_callback(self, sns_client: SnsHelper, event_data: Dict[str, Any], **kwargs):
+        """SNS callback for sending Ray cluster event data to an SNS topic.
+
+        Args:
+            sns_client: Amazon SNS client for publishing to an SNS topic
+            event_data: Ray cluster setup event data. This contains the event name, enum ID, and
+                may also contain additional metadata (i.e. the initialization or setup command used
+                during this setup step)
+            **kwargs: Keyword arguments injected into `_EventSystem.add_callback_handler` before initialization
+        """
+        try:
+            # create a copy of the event data to modify
+            event_dict = copy.deepcopy(event_data)
+            event: RayEvent = event_dict.pop("event")
+            node_context: NodeContext = event_dict.get("node_context", {})
+            sns_topic_arn, params = self.uri, kwargs
+            custom_description = event_dict.get("customDescription")
+            custom_event_name = event_dict.get("customEventName")
+            ray_parent_session_id = event_dict.get("rayParentSessionId")
+            ray_session_id = event_dict.get("raySessionId")
+            message = {
+                **params,
+                "state": event.state,
+                "stateSequence": event.value - 1,  # zero-index sequencing
+                "stateDetailStatus": "SUCCESS",
+                "rayParentSessionId": ray_parent_session_id,
+                "raySessionId": ray_session_id,
+                "timestamp": round(time.time() * 1000),
+            }
+
+            if custom_event_name:
+                message["eventName"] = custom_event_name
+
+            if custom_description:
+                message["stateDetailDescription"] = custom_description
+
+            if node_context:
+                message["rayNodeId"] = node_context["node_id"]
+                message["rayNodeType"] = "HEAD" if node_context["is_head_node"] else "WORKER"
```

Comment:
Let's move the SNS message construction logic into its own method so that it can be easily overridden by any implementation providing a custom SNS message.